elements/pl-matching/pl-matching.py

```diff
@@ -0,0 +1,399 @@
+    } for opt in options_list]
+
+
+def categorize_matches(element, data):
+    """Get provided answers and options from the pl-matching element"""
+    options = {}
+    answers = []
+    index = 0
+
+    # Sort the elements so that pl-answers come first.
+    children = element[:]
+    children.sort(key=lambda child: child.tag, reverse=True)
+
+    for child in children:
+        if child.tag in ['pl-option', 'pl_option']:
+            pl.check_attribs(child, required_attribs=[], optional_attribs=['name'])
+            child_html = pl.inner_html(child)
+            option_name = pl.get_string_attrib(child, 'name', child_html)
+
+            # An option tuple has: index of appearance in the pl-matching element;
+            # the name attribute; and the html content.
+            option_tuple = (index, option_name, child_html)
+            options[option_name] = option_tuple
+            index += 1
+
+        elif child.tag in ['pl-answer', 'pl_answer']:
+            pl.check_attribs(child, required_attribs=['match'], optional_attribs=[])
+            child_html = pl.inner_html(child)
+            match_name = pl.get_string_attrib(child, 'match')
+            if match_name not in options:
+                raise Exception(f'pl-answer "match" attribute of {match_name} does not match any of the pl-option elements.')
+
+            _, match_name, _ = options[match_name]
+
+            # An answer tuple has: the name attribute of the correct matching option; and
+            # the html content.
+            answer_tuple = (match_name, child_html)
+            answers.append(answer_tuple)
+
+    return list(options.values()), answers
+
+
+def prepare(element_html, data):
+    element = lxml.html.fragment_fromstring(element_html)
+
+    required_attribs = ['answers-name']
+    optional_attribs = ['fixed-order', 'number-answers', 'number-options', 'none-of-the-above', 'blank', 'counter-type']
+    pl.check_attribs(element, required_attribs, optional_attribs)
+    name = pl.get_string_attrib(element, 'answers-name')
+    options, answers = categorize_matches(element, data)
+
+    # Choose and randomize the options and answers.
+    # Options are always randomized; answers can be a fixed order.
+    fixed_answer_order = pl.get_boolean_attrib(element, 'fixed-answer-order', FIXED_ANSWER_ORDER_DEFAULT)
+    number_answers = pl.get_integer_attrib(element, 'number-answers', len(answers))
+    number_options = pl.get_integer_attrib(element, 'number-options', len(options))
+    nota = pl.get_boolean_attrib(element, 'none-of-the-above', NOTA_DEFAULT)
+
+    # Shuffle or sample the options.
+    if number_options == len(options):
+        random.shuffle(options)
+    else:
+        options = random.sample(options, number_options)
+        # If we aren't using all options, there is a chance that one answer will be missing its option,
+        # so we force none-of-the-above.
+        nota = True
+    # Then append None of the above if requested.
```

Comment:
Wouldn't it make more sense to select the options associated to the selected answers, plus a random sample of distractor options to complete `number_options`? In that case, if `number_options` is less than the number of options required to fulfill the selected answers, I believe it makes sense to set `nota` to True.

Reply:
Resolved, see comment on PR.